Minimal install

ankurkotwal opened the pull request:
@thefossguy I've built this minimal install that is working well for me.  Some of the changes:

- It really cuts down the number of packages installed.
- Externalises configuration to "image.conf" so you can easily customise the image without diving into the script.
- Locks the root account so it can't be signed into locally or remotely. The user will have to use sudo. This is a best practice and the out of the box experience on Ubuntu for example.

Since you prefer to use doas, I would propose that the chroot-setup.sh script could provide a hook for you to provide an additional script where you could have your customisations for your own needs.

I was building off my own local kernel packages so the checksums in this PR should be ignored. I'm curious to get your thoughts. Feel free to ignore it - I'm happy to maintain a fork.

ankurkotwal committed:
Make image creation configurable. Create minimal install.

ankurkotwal committed:
Lock root account. Use arch1-2 kernel image

ankurkotwal committed:
Add sudo, man-db, man-pages packages. Replace which with command -v. Replace wget with curl. riscv login not working.

ankurkotwal committed:
Remove image if it exists. Fix login issue

ankurkotwal committed:
Merge branch 'master' into minimal-install

thefossguy commented:
Hi!

Before I give feedback on any of the commits, I suggest that you split your work in separate commits. As an example, the first commit (989c2167a05d59cd3e2a302852fe23a179fbb3ee) is introducing a config file to be read from, relevant code is added to ensure that the values aren't empty and changes are made to use those variables.

But the same commit also contains additional changes (changing the packages that are being installed and removing `doas.conf`). These changes are irrelevant _to this commit_. You should separate them to a different commit so it is easy to iterate upon.

thefossguy commented on scripts/install-packages.sh:
```diff
@@ -4,11 +4,8 @@
-# extra pkgs
-#PKGS_TO_INSTALL+=(aria2 base-devel bat btop choose dog dua-cli dust exa fd hyperfine namcap neovim nload procs ripgrep rsync rustup skim tealdeer tre tree unrar unzip wget yt-dlp zsh-autosuggestions zsh-completions zsh-syntax-highlighting)
-
+PKGS_TO_INSTALL=(base networkmanager openssh vim)
```
I see you only kept 4 packages `base networkmanager openssh vim`. I suggest you add at least a few more packages that may look "additional" but are important to diagnose why a fresh new image is not working. Packages like `iputils`, `bind`, `nano` are necessary, not for a functioning system, but to diagnose _why_ it isn't working.

ankurkotwal replied:
More conversation about packages below. Resolving here.

thefossguy commented:
Before I pull in this PR, I need the following addressed.

### 1: The packages chosen for a minimal installation

In my humble opinion, the choice of keeping just 7 packages (`base man-db man-pages networkmanager openssh sudo vim`) isn't good. Yes, that's all that is needed for a very minimal system. But assume if someone has a problem just getting their network setup, diagnosing the issue would require installation of packages like `iputils`, `bind`, etc. It would be hard to do so without a working network connection, you can't even `scp` the files. Similarly, some users might not like Vim, hence `nano` was also in the package list.

I have very carefully chosen the base packages that are, in my opinion, are not just a good base, but are also beneficial in case of a mess up.

Still, if you find any package that is doesn't fit this definition, move it from [line 7](https://github.com/thefossguy/archlinux-visionfive2/blob/master/scripts/install-packages.sh#L7) to [line 10](https://github.com/thefossguy/archlinux-visionfive2/blob/master/scripts/install-packages.sh#L10) (under the comment 'extra packages'). If you do this, I'd also love to hear _why_ you did that. I don't mind constructive criticism ;)


### 2: Checksum mismatch

This is less of an issue but do make sure that you carry the correct checksums in your most recent commit. I am not sure how GH handles this. The only reason why I pointed it out (even though you mentioned it) is because I don't see this reflected in [your repo](https://github.com/ankurkotwal/archlinux-visionfive2/blob/minimal-install/scripts/verify_file_and_checksum.sh#L18).

---

The rest looks good to me. I'll pull in when these issues are addressed. :)

ankurkotwal committed:
Move spl,uboot,kernel,header file URLs and checksums to config. Fail after 3 retries

ankurkotwal committed:
Enable users to provide their own config file.

ankurkotwal commented:
Hey Pratham,
Thank you for the feedback once again. I realise that this PR is a mix of multiple commits that aren't cleanly separate. In part that was because I was working off the risc-v machine and having to copy diffs to my dev machine. You should see this as one large PR rather than individual commits. It changes a lot but few of these changes can be done independently. With that said, I am following your suggested best practice from now.

> Before I pull in this PR, I need the following addressed.
> 
> ### 1: The packages chosen for a minimal installation
> In my humble opinion, the choice of keeping just 7 packages (`base man-db man-pages networkmanager openssh sudo vim`) isn't good. Yes, that's all that is needed for a very minimal system. But assume if someone has a problem just getting their network setup, diagnosing the issue would require installation of packages like `iputils`, `bind`, etc. It would be hard to do so without a working network connection, you can't even `scp` the files. Similarly, some users might not like Vim, hence `nano` was also in the package list.
> 
> I have very carefully chosen the base packages that are, in my opinion, are not just a good base, but are also beneficial in case of a mess up.
> 
> Still, if you find any package that is doesn't fit this definition, move it from [line 7](https://github.com/thefossguy/archlinux-visionfive2/blob/master/scripts/install-packages.sh#L7) to [line 10](https://github.com/thefossguy/archlinux-visionfive2/blob/master/scripts/install-packages.sh#L10) (under the comment 'extra packages'). If you do this, I'd also love to hear _why_ you did that. I don't mind constructive criticism ;)

This is really interesting.  I do agree that there are some packages could be useful for debugging and I'm open to them. Nano makes a lot of sense. But what do iputils and bind offer that aren't in the base package? NetworkManager makes it easy to connect to a wifi. If you have access to Ethernet, this image is plug and play - you can sign in remotely via ssh. Wifi needs a local login to setup wifi and then you're off. 

> 
> ### 2: Checksum mismatch
> This is less of an issue but do make sure that you carry the correct checksums in your most recent commit. I am not sure how GH handles this. The only reason why I pointed it out (even though you mentioned it) is because I don't see this reflected in [your repo](https://github.com/ankurkotwal/archlinux-visionfive2/blob/minimal-install/scripts/verify_file_and_checksum.sh#L18).

I don't understand this comment. What issue are you seeing?

The PR isn't quite ready to merge - I'd like to finalise our list of base packages before merging it in.

Thanks again for the continued help!

thefossguy commented on create-image.sh:
```diff
@@ -1,11 +1,36 @@
+CONFIG="image.conf"
+
+show_usage() {
+    echo "Create Arch Linux image for the Risc-V StarFive VisionFive 2 single board computer\n
```
~~`Risc-V`~~ `RISC-V`

ankurkotwal replied:
Fixed in upcoming commit

thefossguy commented on create-image.sh:
```diff
@@ -1,11 +1,36 @@
+
+
+if [ $# -ne 0 ]; then
+    [ $# -gt 1 ] && (>&2 echo "error: too many arguments" && show_usage && exit 1)
```
Is there any particular reason why we are exiting just because the user provided extra argument(s)? I don't think that we are accepting any arguments other than the config file. So we can just ignore rest of the arguments.

ankurkotwal replied:
Extra arguments mean that the user didn't use the command as expected. We should error on that. Silently ignoring sets the wrong expectation from the user and the behaviour is effectively undefined.

thefossguy replied:
>  Silently ignoring sets the wrong expectation from the user and the behaviour is effectively undefined.

But we aren't even using `$2`, why would that matter? I don't mean to be rude but I legitimately do not understand.

The presence or absence of any additional arguments don't affect the control flow of this script, so imho, it shouldn't matter.

ankurkotwal replied:
> But we aren't even using `$2`, why would that matter? I don't mean to be rude but I legitimately do not understand.
> 
> The presence or absence of any additional arguments don't affect the control flow of this script, so imho, it shouldn't matter.

The presence of unexpected input shouldn't be ignored, it should error. We are validating that the input provided is correct. It's just a best practice.

thefossguy commented:
> This is really interesting. I do agree that there are some packages could be useful for debugging and I'm open to them. Nano makes a lot of sense. But what do iputils and bind offer that aren't in the base package? NetworkManager makes it easy to connect to a wifi. If you have access to Ethernet, this image is plug and play - you can sign in remotely via ssh. Wifi needs a local login to setup wifi and then you're off.

`iputils` provides the `ping` command and `bind` offers the `dig` command.

`archlinux-keyring` is a part of the `base` meta-package but there was a time when the package installation would fail because the Arch ISO was a bit too old so the keyring needed to be installed first. This package can be removed but I kept it just in case such a thing would happen in future. Less frustration for everyone.

Two shells, `bash` and `dash` (extremely **fast** POSIX shell) are installed in the base. An extra shell shouldn't be that much of an issue.

In the file-systems section, packages `btrfs-progs dosfstools e2fsprogs exfatprogs xfsprogs` provide utilities for the btrfs, fat, ext*, exfat and xfs filesystems respectively.

`cronie` is the package that provides `cron`. `findutils` provides `find` and `xargs`. `mlocate` provides `locate` and `updatedb`. `strace` is useful for debugging purposes. I think these are some packages that one might expect to be present in a UNIX-like shell environment.

`lsb-release` is present in the base because some user-space utilities depend on it. `htop` is in the base because almost everyone uses this at one point, better to have this pre-installed.

Some compression related utilities are also present because that is one of the most common task a developer might perform. Either release a tarball or extract someone else's tarball and compile it as a package manager.

If you are interested in knowing more about what a package provides, look at [packages.archlinux.org](https://archlinux.org/packages) and search for a package. On the package's page, you will see a section called 'Package Contents'. Under that, you will see **all** the files that a package contains. Or, you can just add the string '`/files/`' at the end of the package's URL and see these files.

example: [neovim package page](https://archlinux.org/packages/community/x86_64/neovim/) and [files in the neovim package](https://archlinux.org/packages/community/x86_64/neovim/files/)

> > ### 2: Checksum mismatch
> > This is less of an issue but do make sure that you carry the correct checksums in your most recent commit. I am not sure how GH handles this. The only reason why I pointed it out (even though you mentioned it) is because I don't see this reflected in [your repo](https://github.com/ankurkotwal/archlinux-visionfive2/blob/minimal-install/scripts/verify_file_and_checksum.sh#L18).
>
> I don't understand this comment. What issue are you seeing?

If you look at the checksums for the _kernel package_ and the _kernel header package_ in your most recent commit ([[1](https://github.com/ankurkotwal/archlinux-visionfive2/blob/minimal-install/image.conf#L19)] [[2](https://github.com/ankurkotwal/archlinux-visionfive2/blob/minimal-install/image.conf#L21)]), they don't match either the [latest release](https://github.com/thefossguy/linux-starfive-visionfive2/releases/download/v2.10.4-3/CHECKSUMS) nor the checksums in my master branch ([[1](https://github.com/thefossguy/archlinux-visionfive2/blob/master/scripts/verify_file_and_checksum.sh#L18)] [[2](https://github.com/thefossguy/archlinux-visionfive2/blob/master/scripts/verify_file_and_checksum.sh#L22)]).

ankurkotwal committed:
Move package list to config. Validate config.

ankurkotwal committed:
Fix typo in usage screen.

ankurkotwal committed:
Expand package list as per suggestions from @thefossguy. Fix package install script

ankurkotwal committed:
Add zip package

ankurkotwal commented:
@thefossguy I appreciate the breakdown of these packages.
> 
> `iputils` provides the `ping` command and `bind` offers the `dig` command.

Added.

> `archlinux-keyring` is a part of the `base` meta-package but there was a time when the package installation would fail because the Arch ISO was a bit too old so the keyring needed to be installed first. This package can be removed but I kept it just in case such a thing would happen in future. Less frustration for everyone.

Added.
 
> Two shells, `bash` and `dash` (extremely **fast** POSIX shell) are installed in the base. An extra shell shouldn't be that much of an issue.

It's not an issue but in the spirit of this being minimal, I see no harm in folks having to install additional shells themselves. I myself use fish and install it manually.

> In the file-systems section, packages `btrfs-progs dosfstools e2fsprogs exfatprogs xfsprogs` provide utilities for the btrfs, fat, ext*, exfat and xfs filesystems respectively.

Added.

> `cronie` is the package that provides `cron`.

The Arch way to schedule tasks is with systemd. If folks want to use cronie instead, they're welcome to install it themselves and schedule accordingly. I don't feel it needs to be in the base package.

> `findutils` provides `find` and `xargs`.

Added.

> `mlocate` provides `locate` and `updatedb`.

Added.

> `strace` is useful for debugging purposes. I think these are some packages that one might expect to be present in a UNIX-like shell environment.

For developers, yes. Folks can feel free to install this one if they need it but I don't think it needs to be in the base minimal image.

> `lsb-release` is present in the base because some user-space utilities depend on it. `htop` is in the base because almost everyone uses this at one point, better to have this pre-installed.

Added both.

> Some compression related utilities are also present because that is one of the most common task a developer might perform. Either release a tarball or extract someone else's tarball and compile it as a package manager.

`base` already adds `bzip2`, `gzip`, `tar`, `xz` and `zstd`. I've added `zip`. That should cover everything, right?


> If you are interested in knowing more about what a package provides, look at [packages.archlinux.org](https://archlinux.org/packages) and search for a package. On the package's page, you will see a section called 'Package Contents'. Under that, you will see **all** the files that a package contains. Or, you can just add the string '`/files/`' at the end of the package's URL and see these files.

Yup I was aware of this and have been using it. I don't think we should explicitly add packages that are installed by meta packages (i.e. `gzip` in `base`)

> > > ### 2: Checksum mismatch
> > > This is less of an issue but do make sure that you carry the correct checksums in your most recent commit. I am not sure how GH handles this. The only reason why I pointed it out (even though you mentioned it) is because I don't see this reflected in [your repo](https://github.com/ankurkotwal/archlinux-visionfive2/blob/minimal-install/scripts/verify_file_and_checksum.sh#L18).

Yes, these checksums are from the kernels I've built and won't match yours. When you merge, please ensure that the URL and checksum for your repo is matched. Now that you can specify your own config files, we could end up with separate configs on different repos whilst all the actual code remains in sync.

ankurkotwal committed:
Merge branch 'master' into minimal-install

thefossguy commented on scripts/install-packages.sh:
```diff
@@ -4,7 +4,7 @@
 ################################################################################
 
-PKGS_TO_INSTALL="${CONF_PKGS_TO_INSTALL}"
+PKGS_TO_INSTALL=(`echo "${CONF_PKGS_TO_INSTALL}"`)
```
This commit was unnecessary :sweat_smile:

ankurkotwal replied:
Converts a string from the config file into a list. It's needed else pacstrap receives a single argument containing all the packages (which fails) as opposed to a list of packages.

thefossguy replied:
Shouldn't `${CONF_PKGS_TO_INSTALL[@]}` be more appropriate in that case?

thefossguy commented:
I apologize deeply but tracking the changes you have made is getting kind of hard for me, since there are so many commits along with multiple [unrelated] changes in a commit. I repeatedly refer to older commits for changes that were made and to be honest, commits with changes that aren't relevant to _that particular commit_ muddies the water for me. I don't mean to be rude with this but it is what it is and I don't want you to lift the burden of maintaining a fork since I do really appreciate the changes that you have introduced! :)

I suggest you create a _separate_ PR for each of the things that you are improving. I would suggest the following order for sending PRs:

1. Introduce the config file (keep `CONF_PKGS_TO_INSTALL` empty) and source it automatically. Contrary to the commit 8bb258fdce2e467634953a00e11f182a6ef68dd3, the user can just modify the config file and not worry about providing a whole new config file.
3. Move the **base** packages from `scripts/install-packages.sh` to the `CONF_PKGS_TO_INSTALL` array _with the reduced package list_. When you "remove" packages from the base install, move the "removed packages" to somewhere else, preferably [to this array](https://github.com/thefossguy/archlinux-visionfive2/blob/master/scripts/install-packages.sh#L10).
4. Lock root account.
5. Set retry count to 3 (77238e5c8e9ce1c8975479fed4c6f482e0ffb85a)
6. \<Insert more of your ideas here\> :)

---

P.S. You will be able to use [this image](https://pkgbuild.com/~tpowa/archboot/iso/riscv64/latest/) (just like the usual experience you have with Arch Linux on x86 machines) once upstream applies StarFive's patches. So this repository will no longer be relevant after the upstream merge :)

ankurkotwal commented:
Would it help if you looked at this PR as one diff, rather than multiple commits? Think of it as a feature branch.

If not, then I can look into splitting into multiple PRs but it'll take a few days because I'm quite tied up at the moment.

> P.S. You will be able to use [this image](https://pkgbuild.com/~tpowa/archboot/iso/riscv64/latest/) (just like the usual experience you have with Arch Linux on x86 machines) once upstream applies StarFive's patches. So this repository will no longer be relevant after the upstream merge :)

Where's this image from? I like the fact that we have a repo that can build an image from scratch at least until there's something official. Arch still doesn't have official support for aarch64, even though Manjaro does. Also given the current incomplete state of the VisionFive2 kernel, I'm not holding my breath for the upstreaming of those changes. The framebuffer display still doesn't work, let alone anything else. Not saying folks aren't working on it, just that it's unlikely to be available upstream in the near future.

thefossguy commented:
Hi, apologies for the late response. I've been ill since a while. I'll tend to this once I am better. Stay safe :)

ankurkotwal commented:
Really sorry to read about you being unwell. Please don't worry about this and focus on getting better. Get well soon!

ankurkotwal commented:
@thefossguy I hope you're feeling better. When you've recovered and are feeling up to reviewing this, the easiest way I found was to view the PR as a single diff at https://github.com/thefossguy/archlinux-visionfive2/pull/5/files 

You can then see what's changed. I'll be sure to do better separation next time.

thefossguy commented:
Two things before I merge this PR.

1. I still don't know why the rest of packages are **excluded**. I justified my reasons for keeping said packages in the base packages list in a comment somewhere above. You acked them and added only those particular packages to the base pkgs list. Is there a reason why you have not added all the packages that I kept in the base install to your `image.conf` file? (Not arguing with you, but I don't understand what is happening.)

2. The package checksums that are in my master repo don't match the ones stored in your `minimal-install` branch's `image.conf` file. A commit to update those and this issue will be resolved.

---

> I'll be sure to do better separation next time.

Thanks :)

ankurkotwal committed:
Image.conf changed as per @thefossguy's repo

ankurkotwal commented:
> Two things before I merge this PR.
> 
> 1. I still don't know why the rest of packages are **excluded**. I justified my reasons for keeping said packages in the base packages list in a comment somewhere above. You acked them and added only those particular packages to the base pkgs list. Is there a reason why you have not added all the packages that I kept in the base install to your `image.conf` file? (Not arguing with you, but I don't understand what is happening.)

I don't feel like they're needed for a minimal install but happy to go with your list. Since everything is pushed out to image.conf now, its way easier to maintain custom configs. I'll keep a personal minimal config that I prefer, which is different to what you would use. Its just a single config file now. Please see the commit [56ba0fc](https://github.com/thefossguy/archlinux-visionfive2/pull/5/commits/56ba0fc6881104030d4d0dbf846643e61a8443f8) in this PR for that change.

> 2. The package checksums that are in my master repo don't match the ones stored in your `minimal-install` branch's `image.conf` file. A commit to update those and this issue will be resolved.

Sorry that was meant to be cleaned up during the merge. Please see the commit [56ba0fc](https://github.com/thefossguy/archlinux-visionfive2/pull/5/commits/56ba0fc6881104030d4d0dbf846643e61a8443f8) in this PR for that change.

thefossguy commented:
Checked your diff and cloned your fork to test if the image builds (I have OCD). Making this comment to let you know that I have ack-ed your changes and it looks okay to me.

I will either let you know about any issues, if any. Or will pull and close this PR otherwise.

Thank you for your contribution, much appreciated! @ankurkotwal

ankurkotwal commented:
That's great, thanks @thefossguy!

thefossguy commented:
I don't know why this is happening but I am unable to merge your PR. Neither from the Github web UI nor form the `git` cli. :(
Since everything looks good and I was able to build an image successfully, I would love to accept these changes. Do you mind sending me a patch via an email?

Sorry for the issues caused by my mentally exhausted self. Still recovering lol :p

(Closing this as I will merge your changes as soon as I get your patches.)

Again, thanks for your contributions @ankurkotwal :)

thefossguy commented:
I don't know why this is happening but I am unable to merge your PR. Neither from the Github web UI nor form the `git` cli. :(
Since everything looks good and I was able to build an image successfully, I would love to accept these changes. Do you mind sending me a patch via an email?

Sorry for the issues caused by my mentally exhausted self. Still recovering lol :p

Again, thanks for your contributions @ankurkotwal :)

ankurkotwal commented:
Adding patch here for completeness of this PR but will also email it.
[dba8c64d0557803dfbf715f6253fe3fcfc710587-HEAD.patch](https://github.com/thefossguy/archlinux-visionfive2/files/11083043/dba8c64d0557803dfbf715f6253fe3fcfc710587-HEAD.patch)

thefossguy commented:
Your provided patch does not apply. I tried pulling in your branch and diff it against my master branch and generate a patch myself, that too doesn't apply. :(

Do you mind if I take over again? As last time, you will be credited :)

ankurkotwal commented:
Yeah didn't work for me either which is strange. I recreated it as a single squashed commit (instead of concatenated multiple patches) and it works for me now. Could you try again?

EDIT: Attaching the correct file now.
[merge-minimal-install.patch](https://github.com/thefossguy/archlinux-visionfive2/files/11086282/merge-minimal-install.patch)

thefossguy commented:
Merged using GitHub web UI now. Looks like I was unable to merge it earlier because it was an [issue](https://twitter.com/_ediri/status/1640330299522334721) from GH's side.

ankurkotwal commented:
Thanks @thefossguy
